File: 2023/elixir/day02a/day02a.exs
defmodule Day02a do
  @progname "day02a.exs"

  @total_red 12
  @total_green 13
  @total_blue 14

  def usage do
    IO.puts("usage: #{@progname} <file>")
    System.halt(1)
  end

  def process(contents) do
    lines = String.split(contents, "\n", trim: true)
    Enum.reduce(lines, 0, fn line, result ->
      game_part = hd(String.split(line, ": "))
      game_num = String.to_integer(hd(tl(String.split(game_part, " "))))
      draws_part = hd(tl(String.split(line, ": ")))
      valid = Enum.reduce(String.split(draws_part, "; "), true, fn draws, current_valid ->
        Enum.reduce(String.split(draws, ", "), current_valid, fn draw, draw_valid ->
          amount = String.to_integer(hd(String.split(draw, " ")))
          color = hd(tl(String.split(draw, " ")))
          case color do
            "red" -> draw_valid and (amount <= @total_red)
            "green" -> draw_valid and (amount <= @total_green)
            "blue" -> draw_valid and (amount <= @total_blue)
          end
        end)
      end)
      if valid do
        result + game_num
      else
        result
      end
    end)
  end

  def main do
    if length(System.argv()) < 1 do
      usage()
    end

    filename = hd(System.argv())
    {_, contents} = File.read(filename)
    result = process(contents)
    IO.puts("result = #{result}")
  end
end

Day02a.main()

File: 2023/javascript/day02a/day02a.js
const fs = require('fs');

function usage() {
    console.log(`usage: ${process.argv[0]} ${process.argv[1]} <file>`);
    process.exit(1);
}

function processInput(contents) {
    const totalRed = 12;
    const totalGreen = 13;
    const totalBlue = 14;
    let lines = contents.trim().split('\n');
    let result = 0;
    for ( let line of lines ) {
        let gamePart = line.split(': ')[0];
        let gameNum = parseInt(gamePart.split(' ')[1]);
        let valid = true;
        let drawsPart = line.split(': ')[1];
        let draws = drawsPart.split('; ');
        for ( let drawIndex in draws ) {
            draw = draws[drawIndex];
            colorAmounts = draw.split(', ');
            for ( let colorAmountIndex in colorAmounts ) {
                let colorAmount = colorAmounts[colorAmountIndex];
                let amount = parseInt(colorAmount.split(' ')[0]);
                let color = colorAmount.split(' ')[1];
                if ( color == 'red' && amount > totalRed ) {
                    valid = false;
                } else if ( color == 'green' && amount > totalGreen ) {
                    valid = false;
                } else if ( color == 'blue' && amount > totalBlue ) {
                    valid = false;
                }
            }
        }
        if ( valid ) {
            result += gameNum;
        }
    }
    return result;
}

function main() {
    if ( process.argv.length < 3 ) {
        usage();
    }
    let filename = process.argv[2];
    let contents = fs.readFileSync(filename, 'utf8');
    let result = processInput(contents);
    console.log(`result = ${result}`);
}

main()
